src/Algebra/Graph/Acyclic/AdjacencyMap.hs

```diff
@@ -0,0 +1,112 @@
+module Algebra.Graph.Acyclic.AdjacencyMap where
+
+import qualified Algebra.Graph.AdjacencyMap as AM
+import qualified Algebra.Graph.AdjacencyMap.Algorithm as AMA
+import qualified Algebra.Graph.AdjacencyMap.Internal as AMI
+import qualified Algebra.Graph.NonEmpty.AdjacencyMap as NonEmpty
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Trans.State.Strict
+import Data.Map (Map, (!))
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+
+newtype AdjacencyMap a = AAM
+  { aam :: AM.AdjacencyMap a
+  } deriving (Show)
+
+consistent :: Ord a => AdjacencyMap a -> Bool
+consistent (AAM m) = AMI.consistent m && AMA.isAcyclic m
+
+type AdjacencyMapP a = AdjacencyMap (Int, a)
+
+empty :: AdjacencyMap a
+empty = AAM AM.empty
+
+vertex :: a -> AdjacencyMap a
+vertex x = AAM $ AM.vertex x
+
+vertices :: (Ord a) => [a] -> AdjacencyMap a
+vertices xs = AAM $ AM.vertices xs
+
+overlayD ::
+     (Ord a, Ord b)
+  => AdjacencyMap a
+  -> AdjacencyMap b
+  -> AdjacencyMap (Either a b)
+overlayD (AAM a) (AAM b) = AAM (AM.overlay (AM.gmap Left a) (AM.gmap Right b))
+
+connectD ::
+     (Ord a, Ord b)
+  => AdjacencyMap a
+  -> AdjacencyMap b
+  -> AdjacencyMap (Either a b)
+connectD (AAM a) (AAM b) = AAM (AM.connect (AM.gmap Left a) (AM.gmap Right b))
+
+transitiveClosure :: Ord a => AdjacencyMap a -> AdjacencyMap a
+transitiveClosure (AAM x) = AAM (AM.transitiveClosure x)
+
+overlay :: (Ord a) => AdjacencyMap a -> AdjacencyMap a -> AdjacencyMap a
+overlay (AAM x) (AAM y) = AAM (AM.overlay x y)
+
+connect :: (Ord a) => AdjacencyMap a -> AdjacencyMap a -> AdjacencyMap a
+connect (AAM x) (AAM y) =
+  AAM . AMI.AM $
+  Map.unionsWith Set.union $
+  AM.adjacencyMap x :
+  AM.adjacencyMap y :
+  [ Map.fromSet
+      (\k -> Set.filter (< k) . Map.keysSet $ AM.adjacencyMap y)
+      (Map.keysSet $ AM.adjacencyMap x)
+  ]
+
+-- reverse and Map.keys results in an ascending list
+topSort :: (Ord a) => AdjacencyMap a -> [a]
+topSort (AAM (AMI.AM m)) = reverse $ Map.keys m
+
+-- scc
+scc :: (Ord a) => AM.AdjacencyMap a -> AdjacencyMapP (NonEmpty.AdjacencyMap a)
```

Comment:
This doesn't match the requirements: the result should be `AdjacencyMap (NonEmpty.AdjacencyMap a)`.

Reply:
`scc` is a very simple function, something like this:

```haskell
scc m = AAM (AM.scc m)
```

Reply:
It wouldn't be that simple as some edges might break the invariant if the default order is taken into account.

Reply:
Can you give an example? The very definition of `scc` guarantees absence of cycles.

Have a look at the documentation:

https://hackage.haskell.org/package/algebraic-graphs-0.4/docs/Algebra-Graph-AdjacencyMap-Algorithm.html#v:scc

We even have the property `isAcyclic . scc == const True` which is tested in the testsuite. So the result is guaranteed to be acyclic, we do not need any additional requirements.

Reply:
Keeping the invariant in mind an edge could be formed from a lower ordered vertex to a higher one. In that sense, it breaks the invariant. Does not form any cycles though. I've changed it now.